### PR TITLE
Add config.d support

### PR DESCRIPTION
**Pull request by maybe-sybr**
This is an initial look at the work I've been doing related to #135.

### CODE REVIEW

**Comment by maybe-sybr** [comment]
This doesn't have env var support yet for overriding stuff in the on-disk configs. Specifically, I want to be able to provide something like `MEDALLION_BACKEND_MONGO_URI` to override/provide the `.backend.uri` config key which the mongo backend uses. Trying to do that as a general system is annoying - although I've made some progress - but it would be much easier to simply teach the mongo backend itself about such an env-var so it can pull the var in itself. I'm planning to do that in my out-of-tree backend for the moment (using the changeset in #133 ) and then coming back at it when I have a bit more time to think it through.

File: setup.py [hunk 1]
```diff
@@ -48,8 +48,10 @@ def get_long_description():
     keywords="taxii taxii2 server json cti cyber threat intelligence",
     packages=find_packages(exclude=["*.test", "*.test.data"]),
     install_requires=[
+        "environ-config",
```
Comment (maybe-sybr) [inline setup.py:51]: This changeset depends on a PR I have up for this package. If it doesn't get pulled then I have an alternative approach that isn't as nicely declarative but would work fine with the currently released versions of the environs package.

**Comment by maybe-sybr** [comment]
The way I do environment variable config definitions is currently much easier with #133 merged in. I've done that here for the moment.

**Comment by emmanvg** [comment]
It has been merged.

**Comment by maybe-sybr** [comment]
Great, thanks! I'll work on cleaning this up and check in on if my PR in environ-config might be merged. If not, I'll pivot to an alternative approach to get this out of draft state and complete the outline I made in #135

**Comment by maybe-sybr** [comment]
Apologies for the radio silence on this. It's still on my list and a little while ago my changes to `environ-config` were merged which should mean I can clean this up, rebase and de-draft it.

**Comment by maybe-sybr** [comment]
I've now found the time to run through this and prep it for merge from a code perspective. The changeset is freshened on top of current master and I've fixed an issue with config being dropped when not specified in the environment. A couple of questions:
 * what's the expectation around testing? this is new code so I assume you might want some unit tests to cover it
 * @zrush-mitre suggested that this should be documented. I don't see anything config related in the `docs/` tree yet, will you need me to write something up?

File: docker-compose.yml [hunk 2]
```diff
@@ -7,7 +7,10 @@ services:
     build:
       context: .
       dockerfile: docker_utils/Dockerfile
-    command: [sh, -c, "medallion --host 0.0.0.0 docker_config.json"]
+    command: [sh, -c, "medallion --host 0.0.0.0"]
+    env:
+      MEDALLION_BACKEND_MODULE_CLASS: "MongoBackend"
```
Comment (maybe-sybr) [inline docker-compose.yml:12]: I do wish this could just be `MEDALLION_BACKEND: "MongoBackend"` but `environ-config` doesn't allow us to do variable name customisation like that :(

File: setup.py (hunk 1 above)
Comment (maybe-sybr) [reply setup.py:51]: This got merged into `environ-config` 21.1.0. Shall we set this `>=21.1`?

**Review by zrush-mitre — COMMENTED** [review]
Hi @maybe-sybr,

Good job on the PR. Some interesting choices for programming, but functional nonetheless. I'll be doing some testing of it myself in the coming days.

For testing, we definitely want to maximize code coverage, but we don't have a coverage goal or anything. Rather we want to ensure that 1) common use cases are tested, 2) as much of the code is covered by tests (within reason) 3) edge and trouble cases are included and 4) the depth of the tests are increased, as compared to the tests that came before.

4 is really important, since often times when capabilities are added to code, they won't be heavily modified for some time, so we want to make sure that the types of tests that are included are encouraging other PRs to also increase their testing breadth/depth. The config tests before defined a few different uses and edge cases to make sure they had the correct behavior, but for this PR we want to see some config file variances, as well as testing the behaviors of the flowchart you made above to ensure things are and will work correctly.

Some documentation is definitely needed to show what the environment variables are, where the code looks for configs, and how they should be structured. Depending upon how in depth the documentation is, you can either make a new file in the docs folder, or add a straightforward section to the readme. Speaking of...

Make sure stuff like previous tests and previous documentation is modified as needed for your changes. The README may need some changes to show how to call initial servers and whatnot, and other tests may need the same changes.

Thanks for your work, and I'll continue reviewing your code.

File: medallion/config.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    # Load configuration from the default config file or a value passed in
+    if conf_file is not None:
+        conf_file_p = pathlib.Path(conf_file)
+        try:
+            config_data = json.load(conf_file_p.open())
+        except FileNotFoundError as exc:
+            if conf_file is not DEFAULT_CONFFILE:
+                raise exc
+            else:
+                config_data = dict()
```
Comment (maybe-sybr) [inline medallion/config.py:48]: This can be undefined if we skip this when `conf_file is None`. I'm reworking it to be a bit more DRY and fix this bug.

File: medallion/config.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        self.__strip(dictified)
+        return dictified
+
+
+def load_config(conf_file=DEFAULT_CONFFILE, conf_dir=DEFAULT_CONFDIR):
+    # Load configuration from the default config file or a value passed in
+    if conf_file is not None:
+        conf_file_p = pathlib.Path(conf_file)
+        try:
+            config_data = json.load(conf_file_p.open())
```
Comment (maybe-sybr) [inline medallion/config.py:43]: I also need to handle JSON decoding errors and raise something like a `TypeError` if the data we get back is not a JSON object (e.g. array, raw value) since json merge will complain about merging objects and other types and the config we get from `environ` is always going to be an object (ie. `dict`).

**Review by maybe-sybr — COMMENTED** [review]
I'm most of the way through building out the tests for this. I should have something up before I finish working on it today.

File: medallion/config.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            if conf_file is not DEFAULT_CONFFILE:
+                raise exc
+            else:
+                config_data = dict()
+    # Load extra configuration from a config.d style directory at either the
+    # default path or one passed in
+    if conf_dir is not None:
+        conf_dir_p = pathlib.Path(conf_dir)
+        try:
+            for conf_file_p in conf_dir_p.iterdir():
```
Comment (maybe-sybr) [inline medallion/config.py:54]: This also needs to be sorted in some way since `iterdir()` won't do that for us

**Comment by maybe-sybr** [comment]
Hmm, looks like maybe my trick for testing the argument parser doesn't work in 3.7 :/

File: medallion/test/test_startup.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+
+def test_parser_help(capsys):
+    """
+    Confirm that the parser help can be printed with the custom formatter.
+    """
+    parser = medallion.scripts.run._get_argparser()
+    parser.print_help()
+    (out, _) = capsys.readouterr()
+    assert "medallion v" in out
```
Comment (zrush-mitre) [inline medallion/test/test_startup.py:17]: Was this supposed to be the whole text of the help command? I assume "medallion v" is just a random section of it, and that's okay, just seemed odd.

File: medallion/test/config/test_config_loading.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+            )
+            assert config == {}
+
+
+def test_load_config_non_default_file_missing(subtests, tmp_path):
+    """
+    Confirm an exception is raised if a non-default config file is missing.
+    """
+    conf_file = tmp_path / "missing.conf"
+    with pytest.raises(FileNotFoundError, match=repr(str(conf_file))):
```
Comment (zrush-mitre) [inline medallion/test/config/test_config_loading.py:45]: Because of Windows paths, this needs an `re.escape` for the match, like so:

`match=re.escape(repr(str(conf_file))))`

File: medallion/test/config/test_config_loading.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+            )
+            assert config == {}
+
+
+def test_load_config_non_default_dir_missing(tmp_path):
+    """
+    Confirm an exception is raised if a non-default config dir is missing.
+    """
+    conf_dir = tmp_path / "missing.d"
+    with pytest.raises(FileNotFoundError, match=repr(str(conf_dir))):
```
Comment (zrush-mitre) [inline medallion/test/config/test_config_loading.py:71]: Because of Windows paths, this needs an `re.escape` for the match, like so:

`match=re.escape(repr(str(conf_file))))`

File: medallion/test/test_startup.py (hunk 6 above)
Comment (maybe-sybr) [reply medallion/test/test_startup.py:17]: It's just a reliable looking substring from the help text. The current version number follows the `v` at the end here and the test shouldn't care about that. It's mostly to validate that the custom formatter is doing its thing :)

File: medallion/test/config/test_config_loading.py (hunk 7 above)
Comment (maybe-sybr) [reply medallion/test/config/test_config_loading.py:45]: Ah good point! Did you run the tests on a Windows machine? I don't have any dev machines which I could use easily at the moment so if you are able to flush out any other issues which I would have missed because of that it would be helpful :) I'll amend this and the one you tagged below when I get some time tomorrow. Thanks

Comment (zrush-mitre) [reply medallion/test/config/test_config_loading.py:45]: Yeah, made sure to run this with both Unix and Windows. This seems like the only issue though (there may be some issues with environment variables, but I still have to test that).

Comment (maybe-sybr) [reply medallion/test/config/test_config_loading.py]: Made this change and just pushed an amended diff. Let me know if that works on your end. I tried to work out if it would play nicely using a variant of this test and passing in a `WindowsPurePath` but the new code creates a `Path` from whatever we pass in so there's some double escaping happening which makes it annoying for me to validate that the test might function as expected on Windows.

File: medallion/config.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+import pathlib
+
+import attr
+import environ
+import jsonmerge
+
+from . import backends
+
+DEFAULT_CONFFILE = "/etc/medallion.conf"
+DEFAULT_CONFDIR = "/etc/medallion.d/"
```
Comment (zrush-mitre) [inline medallion/config.py:12]: I'm not super familiar with configs and their standard places, but is there a better way we can make this platform agnostic? `/etc` is the obvious choice for Unix environments,  but this does not exist by default on Windows. Is there a file location that would work for both systems?

File: medallion/test/config/test_config_loading.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+    "confdata", ({}, {"foo": "bar"}, {"foo": {"bar": ["baz", ]}}, )
+)
+def test_load_config_json_objects(confdata, subtests, tmp_path):
+    """
+    Confirm that config files get loaded correctly.
+    """
+    conf_file = tmp_path / "medallion.conf"
+    json.dump(confdata, conf_file.open("w"))
+    with subtests.test(msg="Used as config file"):
+        assert m_cfg.load_config(conf_file=conf_file) == confdata
```
Comment (zrush-mitre) [inline medallion/test/config/test_config_loading.py:88]: Because this function checks environment variables, if variables are set when running the tests, it will cause a lot of them to fail. Is there a way to unset environment variables at test execution time, then reset them afterwards?

File: medallion/config.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        except (FileNotFoundError, IsADirectoryError):
+            # Skip missing files since we haven't already exploded which means
+            # they're probably not important enough to care about (or TOCTOU?).
+            # We also don't care about sub-directories of the config dir.
+            pass
+        else:
+            config_data = jsonmerge.merge(config_data, new_data)
+    # Load extra configuration from the environment
+    env_config = MedallionConfig.from_environ().as_dict()
+    config_data = jsonmerge.merge(config_data, env_config)
```
Comment (zrush-mitre) [inline medallion/config.py:89]: While I do like the idea of combining the environment variables and the referenced config files, I can see a possibility of a user having issues when a variable is set or a file in the config folder that they don't know about, and they don't have a good grasp on this part of the config process. It may be worth it to print each config value and where it was retrieved from, either by default or when a CLI argument is set.

File: medallion/scripts/run.py [hunk 12]
```diff
@@ -57,11 +59,49 @@ def _get_argparser():
...
+            Path a directory containing JSON configuration files with names
+            ending in .json or .conf. Defaults to the value of the
+            MEDALLION_CONFDIR environment variable or
+            {medallion.config.DEFAULT_CONFDIR!r}.
+        """),
+    )
+    config_dir_group.add_argument(
+        "--no-conf-dir",
+        action="store_true",
+        help=inspect.cleandoc("""
```
Comment (zrush-mitre) [inline medallion/scripts/run.py:100]: Any reason the cleandoc was used here as opposed to the regular string from the other args?

File: medallion/config.py (hunk 9 above)
Comment (maybe-sybr) [reply medallion/config.py:12]: Ah, good point. I should have twigged on that when you mentioned running the tests on Windows. IMO we'd normally want to use a base directory like `%APPDATA%/medallion/` or something similar. I might quiz a few Windows friends to see if they think that's insane or not.

File: medallion/test/config/test_config_loading.py (hunk 10 above)
Comment (maybe-sybr) [reply medallion/test/config/test_config_loading.py:88]: I believe there is a pytest monkey_patch fixture which might do the job: https://docs.pytest.org/en/6.2.x/monkeypatch.html#monkeypatching-environment-variables

I'll look into some best practice for this - I think patching the `os.environ` dict might be reasonable but if that's not easy/recommended, we could potentially mock `environ_config` to pull from an alternative default source of variables.

File: medallion/config.py (hunk 11 above)
Comment (maybe-sybr) [reply medallion/config.py:89]: That's a great idea, a `--conf-check` option or something could turn on some debugging in this function and then exit before actually running medallion?

File: medallion/scripts/run.py (hunk 12 above)
Comment (maybe-sybr) [reply medallion/scripts/run.py:100]: Huh, I don't even remember doing this. I probably just used it to get rid of the leading indentation and newlines so I could use long triple quoted strings without concatenating them myself

File: medallion/test/config/test_config_loading.py (hunk 10 above)
Comment (zrush-mitre) [reply medallion/test/config/test_config_loading.py:88]: You could also get, delete, then reset them with a pytest fixture, though that's definitely more rough than using something like monkeypatch, if it works.

File: medallion/config.py (hunk 9 above)
Comment (maybe-sybr) [reply medallion/config.py:12]: I've looked into this briefly and came across the appdirs library which seems like it'd be good to use, however it prefers to use XDG paths for Linux which means we'd get a default site config path of `/etc/xdg/medallion/` which would be disruptive for in-place upgrades. As an alternative, I could use the current paths for Linux and use appdirs on other platforms (Windows, macOS). Any thoughts?

Comment (zrush-mitre) [reply medallion/config.py:12]: I'm personally against platform-dependent implementations, though I understand sometimes they are unavoidable. Any thoughts here @clenk?

File: medallion/test/config/test_config_loading.py (hunk 10 above)
Comment (maybe-sybr) [reply medallion/test/config/test_config_loading.py:120]: Done, I sanity checked that the amended diff works with a poisonous envvar by applying the following patch. The previous diff would have all sorts of failures with the same envvar.

```diff
diff --git a/tox.ini b/tox.ini
index fb9be89..2d714ad 100644
--- a/tox.ini
+++ b/tox.ini
@@ -12,6 +12,8 @@ deps =
     responses
     pymongo
     pyjwt
+setenv =
+    MEDALLION_TAXII_MAX_PAGE_SIZE = 1337
 commands =
     pytest --cov=medallion medallion/test/ --cov-report term-missing
 
```

File: medallion/config.py (hunk 11 above)
Comment (maybe-sybr) [reply medallion/config.py:132]: Done in the new commit at the tip of the branch

File: medallion/config.py (hunk 9 above)
Comment (maybe-sybr) [reply medallion/config.py]: I've just pushed a commit which uses the `appdirs` library to provide sane paths for non-Linux platforms. That should give an idea of /a/ solution to the problem. Additionally, now or in future, we could add an opt-in to use XDG paths on Linux as well (ie. use appdirs on that platform too).

Happy for you to strip that commit off if you'd prefer not to use it, and I'll also keep an eye on this thread for input from you and @clenk.

File: setup.py [hunk 13]
```diff
@@ -48,8 +48,11 @@ def get_long_description():
     keywords="taxii taxii2 server json cti cyber threat intelligence",
     packages=find_packages(exclude=["*.test", "*.test.data"]),
     install_requires=[
+        "appdirs>=1.4.4",
```
Comment (maybe-sybr) [inline setup.py:51]: This is chosen because 1.4.4 was the first release they made which was marked as stable. It's still seeing slow development but no releases since mid last year.

File: medallion/config.py (hunk 9 above)
Comment (maybe-sybr) [reply medallion/config.py]: I meant that since the default path would go from `/etc/medallion.conf` in, e.g. the medallion docker image, to `/etc/xdg/medallion/<something>`. But you're right, I forgot that without this diff applied, medallion needs a specific path to a config file specified, and the docker image actually provides that path as `/opt/taxii/medallion.conf`.

I can certainly change this to use `appdirs` on all platforms and adjust any references to the `/etc/medallion{.conf,.d/}` paths.

File: medallion/config.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+if platform.system() in {"Linux"}:
+    # Linux platforms don't use XDG paths (yet)
+    DEFAULT_CONFFILE = "/etc/medallion.conf"
+    DEFAULT_CONFDIR = "/etc/medallion.d/"
+else:
+    # On non-Linux systems we use and app specific config directory with an
+    # optional `config.d/` subdirectory
+    _appdirs = appdirs.AppDirs("medallion", "oasis-open")
+    DEFAULT_CONFFILE = pathlib.Path(_appdirs.site_config_dir)
```
Comment (maybe-sybr) [inline medallion/config.py:22]: This is a directory, not a file! Will fix.

File: medallion/config.py (hunk 9 above)
Comment (maybe-sybr) [reply medallion/config.py]: This is all done in the new diff, lmk what you think

File: medallion/config.py (hunk 14 above)
Comment (maybe-sybr) [reply medallion/config.py]: This is fixed in the new diff

File: medallion/config.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+import jsonmerge
+import packaging.version
+
+from . import backends
+from . import version as _medallion_version
+
+# Find a config directory for the closest compatible version including the
+# current one, or fall back to the current version's one if none are extant
+_curr_version = packaging.version.Version(_medallion_version.__version__)
+_compat_versions = {_curr_version.major, 3, 2}
```
Comment (maybe-sybr) [inline medallion/config.py:19]: I included `_curr_version.major` as well as the explicit value `3` to ensure that v4 would try the v3 path unless some action is taken to avoid that (e.g. if a config schema change happens). I'm not sure what the most sensible dev flow for making a change like that would be, probably something like bumping the `medallion.version.__version__ = "4.0.0.dev0"`

**Comment by clenk** [comment]
Looks like this is why the build is failing, FYI:

> medallion/config.py:5:1: F401 'platform' imported but unused

File: medallion/config.py (hunk 15 above)
Comment (zrush-mitre) [reply medallion/config.py:19]: Would you rather do it this way than not including the `packaging` library and making `_compat_versions` equal to `{_medallion_version, "3", "2"}`? One less package and a little cleaner with the latter.

File: medallion/config.py (hunk 9 above)
Comment (zrush-mitre) [reply medallion/config.py]: Looks really good actually. Would rather pass off the responsibility of configuration directories to that library than us handle it ourselves.

File: medallion/config.py (hunk 15 above)
Comment (maybe-sybr) [reply medallion/config.py:19]: Was that supposed to be `{_medallion_version.__version__, ...}`? That'd cause us to hunt for `/etc/xdg/medallion/X.Y.Z/` rather than just `.../X/` (being the current major version which will not be `3` in future).

The use of `packaging.version.Version` is to get canonical PEP440 version parsing rather than having to rely on something like `__version__.split(".", 1)[0]` which IMO is less semantically obvious.

Happy to go either way on this.

**Comment by maybe-sybr** [comment]
> Looks like this is why the build is failing, FYI:
> 
> > medallion/config.py:5:1: F401 'platform' imported but unused

Ah, remnant of the platform specific code which is gone now. I'll push an amended diff in a moment. Thanks!

File: medallion/config.py (hunk 15 above)
Comment (zrush-mitre) [reply medallion/config.py:18]: I think you might've been right on the first try. Go with that for now and we can always change in the future if it becomes a dependency issue.

**Comment by zrush-mitre** [comment]
Hey @maybe-sybr,

Don't wanna leave you in the dark, so here's a quick update:

@clenk is off for some time, so I'm getting promoted to maintainer. Once I have permissions here, I'm gonna give another last review before merging this in. We appreciate all the effort you put in here!

Thanks!

**Comment by maybe-sybr** [comment]
Thanks for the update, @zrush-mitre. Much appreciated!

File: README.rst [hunk 16]
```diff
@@ -76,17 +76,35 @@ Medallion provides a command-line interface to start the TAXII Server
       --log-level {DEBUG,INFO,WARN,ERROR,CRITICAL}
                             The logging output level for medallion.
 
+      -c CONF_FILE, --conf-file CONF_FILE
+                            Path to a single configuration file. Defaults to the
+                            value of the MEDALLION_CONFFILE environment variable
+                            or /etc/xdg/medallion/3/medallion.conf.
+
+      --conf-dir CONF_DIR   Path a directory containing JSON configuration files
```
Comment (zrush-mitre) [inline README.rst:84]: "Path a directory" should be "Path to a directory"

File: medallion/test/config/test_config_loading.py [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+    """
+    Ensure that config loading emits log messages about paths and data.
+    """
+    confdata = {"foo": "bar"}
+    conf_file = tmp_path / "medallion.conf"
+    json.dump(confdata, conf_file.open("w"))
+    caplog.set_level(logging.DEBUG, logger="medallion.config")
+    with subtests.test(msg="Used as config file"):
+        assert m_cfg.load_config(conf_file=conf_file) == confdata
+        assert f"load configuration from {str(conf_file)!r}" in caplog.text
```
Comment (zrush-mitre) [inline medallion/test/config/test_config_loading.py:158]: This line causes the tests to fail on my Windows machine. The conf_file path has 4 slashes in between directories, while caplog is printing with 2 slashes. I'd imagine doing an extra `re.escape` might work, or whatever the opposite is.

File: medallion/test/config/test_config_loading.py [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+        for record in caplog.records:
+            if '"foo": "bar"' in record.message:
+                break
+        else:
+            raise AssertionError("Config data not seen in log output")
+    caplog.clear()
+    with subtests.test(msg="Used from config dir"):
+        assert m_cfg.load_config(conf_dir=tmp_path) == confdata
+        assert f"and {str(tmp_path)!r}" in caplog.text
```
Comment (zrush-mitre) [inline medallion/test/config/test_config_loading.py:167]: Similar to above, the left side is printing with more slashes than the right on Windows (aren't paths between different OSes wonderful?)

File: medallion/test/test_startup.py [hunk 19]
```diff
@@ -0,0 +1,338 @@
...
+
+def test_confdir(subtests):
+    """
+    Confirm that the --conf-dir option works as expected.
+    """
+    config_path_arg = "/we're/on/the/road/to/nowhere"
+    parser = medallion.scripts.run._get_argparser()
+    with subtests.test(msg="--conf-dir omitted"):
+        args = parser.parse_args([])
+        assert args.conf_dir == medallion.config.DEFAULT_CONFDIR
```
Comment (zrush-mitre) [inline medallion/test/test_startup.py:68]: This test seems to fail when environment variables are set on Windows. `args.conf_dir` comes out as the environment variable I set for `MEDALLION_CONFDIR`.

File: medallion/test/test_startup.py [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+            mock_app.run.assert_called_once_with(**default_medallion_kwargs)
+        mock_app.reset_mock()
+        mock_load_config.reset_mock()
+
+        with subtests.test(msg="CONFIG_PATH provided only"):
+            with mock.patch(
+                "sys.argv", ["ARGV0", config_file_arg]
+            ):
+                medallion.scripts.run.main()
+            mock_load_config.assert_called_once_with(
```
Comment (zrush-mitre) [inline medallion/test/test_startup.py:241]: Fails with Windows system variables set. It expects `medallion.config.DEFAULT_CONFDIR` on the second call, but receives my environment variable instead.

File: medallion/test/test_startup.py [hunk 21]
```diff
@@ -0,0 +1,338 @@
...
+            mock_app.run.assert_called_once_with(**default_medallion_kwargs)
+        mock_app.reset_mock()
+        mock_load_config.reset_mock()
+
+        with subtests.test(msg="--conf-file provided only"):
+            with mock.patch(
+                "sys.argv", ["ARGV0", "--conf-file", config_file_arg]
+            ):
+                medallion.scripts.run.main()
+            mock_load_config.assert_called_once_with(
```
Comment (zrush-mitre) [inline medallion/test/test_startup.py:253]: Same as above

File: medallion/test/config/test_config_loading.py (hunk 17 above)
Comment (maybe-sybr) [reply medallion/test/config/test_config_loading.py:158]: I think this is actually to do with the repr which is being used to format the `conf_file`. That's only being done to add single quotes so I can just include them in the format string myself. I just sanity checked with this, so I am pretty sure my updated diff will play nice on your Windows box :)

```python
>>> backslash = chr(92)
>>> print(backslash)
\
>>> print(repr(backslash))
'\\'
>>> print(f"{backslash!r}")
'\\'
>>> print(f"--{backslash!r}--")
--'\\'--
```

File: README.rst (hunk 16 above)
Comment (maybe-sybr) [reply README.rst:84]: fixed here and also in the actual parser help text

File: medallion/test/config/test_config_loading.py (hunk 18 above)
Comment (maybe-sybr) [reply medallion/test/config/test_config_loading.py:167]: As above, please resolve if it's fixed!

File: medallion/test/test_startup.py (hunk 19 above)
Comment (maybe-sybr) [reply medallion/test/test_startup.py:68]: Urgh, I made a bunch of patches for `os.environ` but must have forgotten to do that for these ones. TBH I think a module level fixture for the two new test modules would be more sensible so that we don't need to think about it in future changes.

Comment (maybe-sybr) [reply medallion/test/test_startup.py:68]: Okay, I believe this (and the variants commented below) should be fixed now

**Comment by maybe-sybr** [comment]
@zrush-mitre - I've fixed up those issues (will need confirmation on the windows path rendering) and rebased on top of master. LMK if there's anything I missed :)

**Comment by zrush-mitre** [comment]
@maybe-sybr Looking good! LGTM!

**Comment by maybe-sybr** [comment]
Thanks @zrush-mitre :)
